File: src/main/java/dao/AbstractDao.java
package dao;

import util.jpa.Transactional;

import javax.inject.Inject;
import javax.persistence.EntityManager;
import java.lang.reflect.ParameterizedType;
import java.util.List;

public abstract class AbstractDao<T> implements Dao<T> {
	private final Class<T> entity;

	@Inject
	private EntityManager manager;

	@SuppressWarnings("unchecked")
	AbstractDao() {
		Class<T> genericClass = (Class<T>) getClass().getGenericSuperclass();
		ParameterizedType parameterized = (ParameterizedType) genericClass.getGenericSuperclass();
		entity = (Class<T>) parameterized.getActualTypeArguments()[0];
	}

	public AbstractDao(final Class<T> persistentClass) {
		super();
		this.entity = persistentClass;
	}

	protected Class<T> getEntity() {
		return entity;
	}

	@Override
	public EntityManager getEntityManager() {
		return this.manager;
	}

	@Override
	public List<T> listarTodos() {
		return manager.createQuery("from " + entity.getSimpleName(), entity)
				.getResultList();
	}

	@Override
	@Transactional
	public T salvar(T entity) {
		manager.persist(entity);
		entity = manager.merge(entity);

		return entity;
	}

	@Override
	@Transactional
	public T atualizar(T entity) {
		return manager.merge(entity);
	}

	@Override
	@Transactional
	public T encontraPeloId(T entity) {
		return manager.find(this.entity, entity);
	}

}

File: src/main/java/dao/AbstractDao.kt
package dao

import org.hibernate.Session
import util.jpa.Transactional
import javax.inject.Inject
import javax.persistence.EntityManager

abstract class AbstractDao<T>(persistentClass: Class<T>) : Dao<T> {

    @Inject
    lateinit var manager: EntityManager

    private var entity: Class<T> = persistentClass

    override fun getEntityManager(): EntityManager {
        return this.manager
    }

    override fun getSession(): Session {
        return manager.delegate as Session
    }

    override fun listarTodos(): List<T> {
        return manager.createQuery("from " + entity.simpleName, entity)?.resultList as List<T>
    }

    @Transactional
    override fun salvar(entity: T): T {
        var newEntity = entity
        manager.persist(entity)
        newEntity = manager.merge(newEntity)

        return newEntity
    }

    @Transactional
    override fun atualizar(entity: T): T {
        return manager.merge(entity)!!
    }

    @Transactional
    override fun encontraPeloId(entity: T): T {
        return manager.find(this.entity, entity)!!
    }
}